Rewrite Markdownz with remark

srallen opened the pull request:
Package: lib-react-components

Describe your changes:
We need a way to render markdown. This adds Markdownz rewritten using remark. Reasoning will be in ADR-11. ADR-11 will note whether or not we have parity and what changes are being made. 

TO DO:

- [x] Check if HTML5 video works
- [x] Check image sizing markup works
- [x] Docs
- [x] Tests

# Review Checklist

## General

- [ ] Are the tests passing locally and on Travis?
- [ ] Is the documentation up to date?
- [ ] Is the changelog updated?

## Apps

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `rm -rf node_modules/ && npm install` and app works as expected?

srallen commented:
This is ready for review. Check out #329 for an explanation of the rewrite.

eatyourgreens commented on packages/lib-react-components/src/Markdownz/Markdownz.spec.js:
```diff
@@ -11,4 +13,222 @@ describe('<Markdownz />', function () {
+    })
+
+    describe('when the media is an image', function () {
+      const imagePropsMock = { src: 'https://via.placeholder.com/350x350', alt: 'image-alt-text', children: undefined }
```
Real alt text will have spaces, and should end in a full stop, so I'd use a sentence for the mock text here.

eatyourgreens commented on packages/lib-react-components/src/Markdownz/Markdownz.spec.js:
```diff
@@ -11,4 +13,222 @@ describe('<Markdownz />', function () {
+
+    describe('when the media is an image', function () {
+      const imagePropsMock = { src: 'https://via.placeholder.com/350x350', alt: 'image-alt-text', children: undefined }
+      const imagePropsMockWithSize = { src: 'https://via.placeholder.com/350x350', alt: 'image-alt-text =100x100', children: undefined }
```
Ditto re. image alt text here.

eatyourgreens commented on packages/lib-react-components/src/Markdownz/Markdownz.spec.js:
```diff
@@ -11,4 +13,222 @@ describe('<Markdownz />', function () {
+      it('should remove the width and height declaration from the alt text before setting it on the rendered Image', function () {
+        wrapper.instance().renderMedia(imagePropsMockWithSize)
+        const returnedValue = renderMediaSpy.returnValues[0]
+        expect(returnedValue.props.alt).to.equal('image-alt-text')
```
Expected alt text here should be changed too. Maybe store the alt text string in a constant rather than repeating it throughout the test function.

srallen commented:
I just noticed that I forgot to finish writing tests for returning the Video or audio element from `renderMedia`

srallen commented:
@eatyourgreens thanks for the suggestions. Tests are updated including adding tests for returning video or audio that I forgot to do.

srallen commented:
@eatyourgreens are the updates here ok? @rogerhutchings do you have any comments? Thanks!

eatyourgreens commented:
The storybook looks good, on a quick pass through, and tests are all passing. The superscript and subscript examples aren't laid out as expected in Chrome, but the HTML is correct so I'll put that down to something weird in the browser or the storybook styles.

I'm happy to merge this so that it can be used in the classifier.

eatyourgreens commented:
NB. I haven't gone through all 1,600 lines of code but I can do that tomorrow if you need a detailed review.

srallen commented:
I think a lot of that is the package-lock.json change and another chunk just the storybook demo. The detailed review really just needs to be the Markdownz component and custom ping plugin (which just needs sanity check; I don't necessarily expect you to read up on all of the unifed.js API which I don't know in detail yet myself).

eatyourgreens commented:
No problem. I'll read through those first thing tomorrow morning.

eatyourgreens commented on packages/lib-react-components/src/Markdownz/Markdownz.spec.js:
```diff
@@ -11,4 +13,257 @@ describe('<Markdownz />', function () {
+      expect(url).to.equal('')
+    })
+
+    it('should return an empty string if the resource is falsy, but the mention symbol is', function () {
```
'mention symbol is present'?

eatyourgreens commented on packages/lib-react-components/src/Markdownz/Markdownz.spec.js:
```diff
@@ -11,4 +13,257 @@ describe('<Markdownz />', function () {
+      it('should return the expected url with a search query param', function () {
+        const resource = 'tigers'
+        const baseURL = ''
+        const expectedReturnValue = `${baseURL}/talk/search?query=${resource}`
```
This URL runs a text search for `resource`, which is a long-standing bug  that I think has been bothering volunteers. Eg. this URL returns comments that don't include '#transit'.
https://www.zooniverse.org/projects/nora-dot-eisner/planet-hunters-tess/talk/search?query=transit

Can we change Markdownz to add the # character (%23) to the search?
https://www.zooniverse.org/projects/nora-dot-eisner/planet-hunters-tess/talk/search?query=%23transit

srallen replied:
Thanks, I was unaware that was a bug. Returning this URL is still the current functionality of Markdownz, so we should decide if we want to fix it there too since Talk probably won't be updated for a while.

eatyourgreens replied:
I think hashtags have been broken in Talk since it first launched. We can fix it in the existing code if it's a small patch. Otherwise, this could be one reason to migrate Talk across to the new component.

eatyourgreens commented on packages/lib-react-components/src/Markdownz/Markdownz.spec.js:
```diff
@@ -11,4 +13,257 @@ describe('<Markdownz />', function () {
+      it('should return a url with props.baseURI if defined', function () {
+        const resource = 'tigers'
+        const baseURI = 'https://classrooms.zooniverse.org'
+        const expectedReturnValue = `${baseURI}/talk/search?query=${resource}`
```
Ditto re including the # character in the search term here.

eatyourgreens commented on packages/lib-react-components/src/Markdownz/Markdownz.spec.js:
```diff
@@ -11,4 +13,257 @@ describe('<Markdownz />', function () {
+
+    describe('when the symbol is @', function () {
+      it('should return a url for the user profile', function () {
+        const resource = 'srallen'
```
Should we test for usernames that include non-alphabetic characters? I think `am.zooni` has had problems being @-mentioned in the past.

eatyourgreens commented on packages/lib-react-components/src/Markdownz/Markdownz.spec.js:
```diff
@@ -11,4 +13,257 @@ describe('<Markdownz />', function () {
+        wrapper.setProps({ projectSlug: '' })
+      })
+
+      it('should return a url with props.baseURI if defined', function () {
```
You're repeating the same test functions for each symbol, so consider abstracting the contents of these `describe` blocks into a function that checks the three expectations (query param, project slug and base URI) for a given symbol and resource.

srallen replied:
There's a slight variation in a couple of the tests, but I can do this if you feel it's important.

eatyourgreens replied:
It's no big deal. I thought it would save on some maintenance if test functions were being repeated.

eatyourgreens commented on packages/lib-react-components/src/Markdownz/Markdownz.spec.js:
```diff
@@ -11,4 +13,257 @@ describe('<Markdownz />', function () {
+    expect(wrapper.html().includes('Zooniverse')).to.be.true
+  })
+
+  describe('#buildResourceURL', function () {
```
Testing the internal implementation of a component like this is brittle. If the internal methods ever change, these tests will need to be rewritten and it will be hard to tell if the new implementation conforms to the original behaviour.

Ideally, the Markdownz tests would compare some input markdown with some output HTML (or JSX?) without really caring about the details of how the markdown is parsed.

srallen replied:
I see your point of having an overall test of the Markdownz component comparing markdown to outputted jsx for the customizations. The storybook is basically a manual way of doing that, so we just need an automated test too.

eatyourgreens commented on packages/lib-react-components/src/Markdownz/Markdownz.spec.js:
```diff
@@ -11,4 +13,257 @@ describe('<Markdownz />', function () {
+      })
+    })
+
+    describe('when the media is a video', function () {
```
'when the media is audio'.

eatyourgreens commented on packages/lib-react-components/src/Markdownz/Markdownz.spec.js:
```diff
@@ -11,4 +13,257 @@ describe('<Markdownz />', function () {
+    })
+
+    describe('when the media is a video', function () {
+      const audioPropsMock = { src: 'https://panoptes-uploads.zooniverse.org/production/subject_location/1c93591f-5d7e-4129-a6da-a65419b88048.mpga', alt: 'Street noise', children: undefined }
```
How is `alt` handled for audio elements? It's not tested here.

srallen replied:
It is not. I think the original markdown-it plugin we were using wasn't setting it since I took a look at that for reference. I could set the alt anyway for audio.

eatyourgreens replied:
I think Grommet sets it as `aria-label` on the video element, so we could copy that behaviour for audio.

eatyourgreens commented on packages/lib-react-components/src/Markdownz/lib/ping.js:
```diff
@@ -0,0 +1,134 @@
+// This is basically a fork of https://github.com/zestedesavoir/zmarkdown/tree/master/packages/remark-ping
+// The changes made here make the plugin more flexible for supporting hashtags or other custom mentions
```
Can you be specific about what you've changed?

eatyourgreens replied:
Looking through the code, if you could comment the lines that you changed, I think that would be sufficient.

eatyourgreens commented on packages/lib-react-components/src/Markdownz/Markdownz.js:
```diff
@@ -1,34 +1,146 @@
+        ping: (resource, symbol) => this.shouldResourceBeLinkable(resource, symbol), // We could support passing in a prop to call a function here
+        pingSymbols: [at, hashtag, subjectSymbol],
+        resourceURL: (resource, symbol) => this.buildResourceURL(resource, symbol),
+        matchRegex: /(@[\w]+)|(^#[\w]+$)|(^\^S[0-9]+$)/
```
Maybe use `pingSymbols` to build this expression, so that you don't have to edit symbols in two places in the code.

eatyourgreens replied:
`@[\w]+` will miss valid usernames that have non-alphabetic characters in them.

srallen replied:
I'm working on this, but bizarrely updating this is breaking other remark plugins; specifically the remark-emoji plugin. I think what this is is actually a bug in the ping plugin itself returning an undefined value for a hast node and then the emoji plugin is breaking because it isn't doing an existential check on the nodes. I'm going to try to fix this. The use of the regex in the  ping plugin seems to be brittle, but I'm not sure how else to do this. The original plugin used regex as well as our custom plugin with markdown-it.

srallen replied:
The issue isn't the regex for the username itself, it's the check for the preceding space: `(^|\s)`. Removing that, the parser runs fine. Adding it, then remark-emoji breaks, I think because with it remark-ping is now returning something as undefined when it shouldn't.

eatyourgreens replied:
I don't think this needs to test for a leading space. I think the regex should be
`matchRegex: /(@[\w\-\.]+)|(^#[\w]+$)|(^\^S[0-9]+$)/`

The pattern seems to be the match to apply to each token, after the tokenizer has done its thing.

eatyourgreens replied:
We could test that with the existing regex by running a test for `sometext@srallen` and seeing if that skips your username when the preceding space isn't present.

srallen replied:
I need to double check, but I think it'll miss the username mention if in the middle of a sentence without checking for a space.

eatyourgreens replied:
Add that to the tests first, then fix them if they break.

eatyourgreens replied:
Looking at the original ping code, they don't search for leading space either.
https://github.com/zestedesavoir/zmarkdown/blob/44a120b1d67013b2d19f017448877490e5615a1f/packages/remark-ping/src/index.js#L9

I could be wrong but I think the regex is matched against parsed tokens, not the full markdown text.

srallen replied:
Unfortunately I can't use tests to help fix this since the tests won't run because the parser is throwing.

srallen replied:
Actually I can write tests with the check for the space removed from the regex to make sure it's correctly parsing for at-mentions.

srallen commented on packages/lib-react-components/src/Markdownz/lib/ping.js:
```diff
@@ -0,0 +1,134 @@
+        }, {
+          type: 'emphasis',
+          data: {
+            hName: 'span',
```
I think it's wrapping the em with a span with an added class. Let me double check the original default behavior of the plugin.

srallen commented:
@eatyourgreens I think I've addressed all of your comments. The regex is reworked so that users like `@am.zooni` can be mentioned and specifically mentioned in a sentence `I am pinging @am.zooni.` so that the middle period is grabbed, but the final one is not. There is a test for this now.

eatyourgreens commented on packages/lib-react-components/src/Markdownz/Markdownz.js:
```diff
@@ -1,34 +1,152 @@
+    }
+    if (mimeType && mimeType.includes('audio')) {
+      // Grommet doesn't have an audio component right now
+      return <audio aria-title={alt} controls preload='metadata' src={src}><Paragraph>This audio file is not supported</Paragraph></audio>
```
VoiceOver reads the audio element as 'audio group' in the storybook. I think `aria-title` here should be [`aria-label`](https://www.w3.org/TR/wai-aria/#aria-label).

eatyourgreens commented on packages/lib-react-components/src/Markdownz/Markdownz.js:
```diff
@@ -1,34 +1,152 @@
+  }
+}
+
+// (^|\s)@([\w\-.]+\b)|(^|\s)#([-\w\d]{3,40})|(^|\s)\^S([0-9]+)
```
Leftover comment from the old code.

eatyourgreens commented on packages/lib-react-components/src/Markdownz/Markdownz.js:
```diff
@@ -1,34 +1,152 @@
+        ping: (resource, symbol) => this.shouldResourceBeLinkable(resource, symbol), // We could support passing in a prop to call a function here
+        pingSymbols: [at, hashtag, subjectSymbol],
+        resourceURL: (resource, symbol) => this.buildResourceURL(resource, symbol),
+        matchRegex: /\@([\w\-.]+)(?<!\.)|#([-\w\d]{3,40})|(\^S[0-9]+)/
```
`\@([\w\-.]+)\b` might work better than the `(?<!\.)` negative lookbehind here. It will catch all word boundaries at the end of a username. 

https://regexr.com/ says `?<!` may not be supported in all  browsers, but doesn't say which browsers do support it, so `\b` might be safer for cross-browser support too.

srallen replied:
Burned again on trying to use something new. Firefox and Edge don't support it yet: http://kangax.github.io/compat-table/es2016plus/ Thanks for that catch.

eatyourgreens commented on packages/lib-react-components/src/Markdownz/Markdownz.spec.js:
```diff
@@ -1,14 +1,295 @@
+  })
+
+  it('correctly parses a user at-mention in a sentence', function () {
+    const sentence = `I would like to ping @am.zooni.`
```
You can test your username regex more thoroughly by running this test for a set of usernames by looping through an array of test cases.
```javascript
describe('at-mentions', function () {
    ['team', 'am.zooni', 'a-user'].forEach(function (username) {
      describe(username, function () {
        it('should be recognised at the end of a sentence', function () {
          const sentence = `I would like to ping @${username}.`
          wrapper = shallow(<Markdownz>{sentence}</Markdownz>)
          const pingAnchor = wrapper.find({ href: `/users/${username}` })
          expect(pingAnchor).to.have.lengthOf(1)
          expect(pingAnchor.text()).to.equal(`@${username}`)
        })
      })
    })
 })
```

When I run that set of tests, `@am.zooni` and `@a-user` pass but `@team` fails. I'm not sure why.

Specs for the beginning and middle of a sentence should be tested too. I think the regex will catch them, but tests would check that.

eatyourgreens replied:
I'm an idiot. `@team` is listed in the code as an exception. `@srallen` passes. I'll push a commit.

eatyourgreens replied:
Resolved by 7953dcb.

srallen commented:
@eatyourgreens ok I think this is ready for final review now. Keep in mind that I intend to replace the react component that is rendered for `img` with the new Media component I've submitted in my other PR.

eatyourgreens commented:
Could the Media component be used for video and audio too? It does mime type checking.

srallen commented:
Yes, it will be used for audio and video, so in essence it will replace the `renderMedia` function I initially wrote here.

eatyourgreens commented:
I think that, on its own, makes the switch to Remark worthwhile.

srallen commented on packages/lib-react-components/src/Markdownz/Markdownz.js:
```diff
@@ -1,34 +1,150 @@
+      .use(ping, {
+        ping: (resource, symbol) => this.shouldResourceBeLinkable(resource, symbol), // We could support passing in a prop to call a function here
+        pingSymbols: [at, hashtag, subjectSymbol],
+        resourceURL: (resource, symbol) => this.buildResourceURL(resource, symbol),
```
I think it makes it very clear that the plugin is returning something that we can call a function on here, specifically what's being plucked out by the regex as the resource and its symbol.
